base_layer/wallet/src/storage/sqlite_db/wallet.rs

```
@@ -438,92 +442,136 @@ impl WalletBackend for WalletSqliteDatabase {
+                let argon2_params = Argon2Parameters::from_version(Some(version))?;
+
+                // Derive a secondary key from the existing passphrase and salt
+                let secondary_key = derive_secondary_key(existing, argon2_params.clone(), &secondary_key_salt)?;
+
+                // Attempt to decrypt the encrypted main key
+                let main_key = decrypt_main_key(&secondary_key, &encrypted_main_key, argon2_params.id)?;
+
+                // Now use the most recent version
+                let new_argon2_params = Argon2Parameters::from_version(None)?;
+
+                // Derive a new secondary key from the new passphrase and a fresh salt
+                let new_secondary_key_salt = SaltString::generate(&mut OsRng);
+                let new_secondary_key =
+                    derive_secondary_key(new, new_argon2_params.clone(), &new_secondary_key_salt.to_string())?;
+
+                // Encrypt the main key with the new secondary key
+                let new_encrypted_main_key = encrypt_main_key(&new_secondary_key, &main_key, new_argon2_params.id)?;
+
+                // Store the new secondary key version, secondary key salt, and encrypted main key
+                conn.transaction::<_, Error, _>(|| {
+                    // If any operation fails, trigger a rollback
+                    WalletSettingSql::new(DbKey::SecondaryKeyVersion, new_argon2_params.id.to_string())
+                        .set(&conn)
+                        .map_err(|_| Error::RollbackTransaction)?;
+                    WalletSettingSql::new(DbKey::SecondaryKeySalt, new_secondary_key_salt.to_string())
+                        .set(&conn)
+                        .map_err(|_| Error::RollbackTransaction)?;
+                    WalletSettingSql::new(DbKey::EncryptedMainKey, new_encrypted_main_key.to_hex())
+                        .set(&conn)
+                        .map_err(|_| Error::RollbackTransaction)?;
+
```

An additional _defence-in-depth_ approach here would be, before the transaction closure, to read the 3 values and verify that they were changed as instructed and if not trigger the rollback. (_Needless to say, for this reading the 3 values could be DRYed up._) This would be in line with the diesel examples.

---

Is it possible that such a write operation could apparently succeed without actually writing the expected data?

---

This should not be necessary, the database should throw an error if the set command did not do as expected.